### PR TITLE
out_http: adding support for intermediate certificates

### PR DESCRIPTION
**Pull request by tyarimi**
Signed-off-by: Tzach Yarimi <tyarimi@salesforce.com>

**Which issue(s) this PR fixes**: 
Fixes #

**What this PR does / why we need it**: 
Add support for sending a client certificate chain for mutual TLS authentication in the `out_http` plugin.

**Docs Changes**:

**Release Note**:

### CODE REVIEW

**Comment by repeatedly** [comment]
@nurse This patch modifies Net::HTTP directly. Do you know this approach is safe or not?

**Comment by tyarimi** [comment]
Hi folks. Any chance of merging this PR? We have been running it in production for some time now and it works fine.

**Comment by nurse** [comment]
Of course it's not safe. It should use public APIs.

**Comment by tyarimi** [comment]
@nurse `extra_chain_cert` was added in a later version of ruby. Is there any alternative?

**Comment by nurse** [comment]
Well, it should add more comment for example referring https://github.com/ruby/ruby/commit/31af0dafba6d3769d2a39617c0dddedb97883712#diff-8c2ab8e0fb4f052e1d95ab6334e192c1
Also it should have a condition to avoid the monkey patch runs even if latest ruby.

**Comment by tyarimi** [comment]
@nurse I added the comment, but I'm not sure how to implement the conditional patching (not a ruby expert). Can you please help with that?

File: lib/fluent/plugin/out_http.rb [hunk 1]
```diff
@@ -171,7 +180,15 @@ def setup_http_option
         end
         if @tls_client_cert_path
           raise Fluent::ConfigError, "tls_client_cert_path is wrong: #{@tls_client_cert_path}" unless File.file?(@tls_client_cert_path)
-          opt[:cert] = OpenSSL::X509::Certificate.new(File.read(@tls_client_cert_path))
+
+          bundle = File.read(@tls_client_cert_path)
+          bundle_certs = bundle.scan(/-----BEGIN CERTIFICATE-----(?:.|\n)+?-----END CERTIFICATE-----/)
```
Comment (nobu) [inline lib/fluent/plugin/out_http.rb:185]: This regexp seems to match also the middle of a line, is this intentional?

File: lib/fluent/plugin/out_http.rb [hunk 2]
```diff
@@ -21,6 +21,15 @@
...
 require 'fluent/plugin_helper/socket'
 
+# patch Net::HTTP to support extra_chain_cert which was added in Ruby feature #9758.
+# see: https://github.com/ruby/ruby/commit/31af0dafba6d3769d2a39617c0dddedb97883712
+class Net::HTTP
+  SSL_IVNAMES << :@extra_chain_cert unless SSL_IVNAMES.include?(:@extra_chain_cert)
+  SSL_ATTRIBUTES << :extra_chain_cert unless SSL_ATTRIBUTES.include?(:extra_chain_cert)
+
+  attr_accessor :extra_chain_cert
+end
```
Comment (nurse) [inline lib/fluent/plugin/out_http.rb:31]: With assuming SSL_ATTRIBUTES include :extra_chain_cert if SSL_IVNAMES include :@ extra_chain_cert, single condition can enclose the definitions.
```suggestion
unless SSL_IVNAMES.include?(:@extra_chain_cert)
  class Net::HTTP
    SSL_IVNAMES << :@extra_chain_cert 
    SSL_ATTRIBUTES << :extra_chain_cert
    attr_accessor :extra_chain_cert
  end
end
```

**Review by nurse — CHANGES_REQUESTED** [review]

**Comment by tyarimi** [comment]
Closing this PR in favor of #3146 (switching head branch)

**Pull request closed without merging**